src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -0,0 +1,193 @@
+    extent_types_t type,
+    const CachedExtentRef&,
+    segment_off_t length) = 0;
+
+  virtual scan_device_ertr::future<> scan_device() = 0;
+  virtual ~ExtentAllocator() {};
+};
+
+using ExtentAllocatorRef = std::unique_ptr<ExtentAllocator>;
+
+class SegmentedAllocator : public ExtentAllocator {
+public:
+  using alloc_extent_ertr = SegmentedRewriter::alloc_extent_ertr;
+  SegmentedAllocator(
+    SegmentProvider& sp,
+    SegmentManager& sm,
+    Cache& cache)
+    : segment_provider(sp),
+      segment_manager(sm),
+      cache(cache)
+  {}
+
+  CachedExtentRef alloc_ool_extent(
+    extent_types_t type,
+    const CachedExtentRef& old_extent,
+    segment_off_t length) final;
+
+
+  template <typename T>
+  TCachedExtentRef<T> alloc_ool_extent(
+    const CachedExtentRef& old_extent,
+    segment_off_t length)
+  {
+    auto ltl = calc_target_merit(old_extent);
+    auto iter = writers.find(ltl);
+    if (iter == writers.end()) {
+      iter = writers.emplace(
+        ltl,
+        std::make_unique<SegmentedRewriter>(
+          segment_provider,
+          segment_manager)).first;
+    }
+    auto& writer = iter->second;
+
+    auto nextent = cache.alloc_new_extent<T>(length);
+    nextent->set_paddr({ZERO_SEG_ID, fake_paddr_off});
+    fake_paddr_off += length;
+    nextent->extent_writer = writer.get();
+    return nextent;
+  }
+
+  ExtentAllocator::scan_device_ertr::future<> scan_device() final;
+private:
+  int64_t fake_paddr_off = 0;
+
+  merit_t calc_target_merit(const CachedExtentRef&) const {
+    if (!writers.size()) {
+      return 0;
+    }
+    return std::rand() % writers.size();
+  }
+
+  SegmentProvider& segment_provider;
+  SegmentManager& segment_manager;
+  std::map<merit_t, SegmentedRewriterRef> writers;
+  Cache& cache;
+};
+
+class ExtentPlacementManager {
+public:
+  using alloc_extent_ertr = SegmentedRewriter::alloc_extent_ertr;
+  using scan_devices_ertr = SegmentedAllocator::scan_device_ertr;
+  ExtentPlacementManager(Cache& cache) : cache(cache) {}
+
+  CachedExtentRef alloc_new_extent_by_type(
+    extent_types_t type,
+    const CachedExtentRef& old_extent,
+    segment_off_t length) {
+    auto h = calc_target_merit(old_extent);
+    auto iter = extent_allocators.find(h);
+    auto& allocator = iter->second;
+
+    return allocator->alloc_ool_extent(type, old_extent, length);
+  }
+
+  void add_allocator(merit_t hl, ExtentAllocatorRef&& allocator) {
+    auto [it, inserted] = extent_allocators.emplace(hl, std::move(allocator));
+    assert(inserted);
+  }
+
+  scan_devices_ertr::future<> scan_devices() {
+    return crimson::do_for_each(
+      extent_allocators,
+      [this](auto& entry) {
+      auto& extent_allocator = entry.second;
+      return extent_allocator->scan_device();
+    });
+  }
+
+protected:
+  merit_t calc_target_merit(const CachedExtentRef& extent) const {
+    assert(extent_allocators.size());
+    return std::rand() % extent_allocators.size();
```

Comment:
So merit_t is an index?  I don't understand.

Reply:
Um... allocators/writers used to be indexed by heat_level/lifetime_level. But, after discussion with @tchaikov, we think it may be better to generialize the interface a little bit, as there's still a possibility that we may not allocate extents by heat_level/lifetime_level at all(although the chance of this is relatively low). merit_t is just a placeholder for the index type, it's probable that we decide it should be a template parameter eventually. We defined it to be uint64_t just for simplicity for now. And because we can't determine the content of the hint that should be passed to this target calculator and it seems to be highly probable that the hint would come from the fields of the old extents that are being rewritten, so we passed the const reference of the old extents to the calculator.

Reply:
I think merit_t here confuses this PR and doesn't really do anything to clarify future additions to seastore.  Please remove it and just have a single segment for now.  We can address parameterizing segments by heat/age in a future PR.

Reply:
will do.

Reply:
@athanatos Hi, sam. I dropped merit_t and calc_target_merit(), but I templated the allocator index type, hint type and target calculator type. I think this should provide a clear interface for future development. And during the testing, I found that bugs tend to be more easily popped up if there are multiple ool writers, so I kept the multiple segments support, please take a look again.